backend/drm/drm.c

```diff
@@ -0,0 +1,597 @@
+		wlr_log(L_ERROR, "Failed to create EGL display: %s", egl_error());
+		goto error_gbm;
+	}
+
+	if (eglInitialize(renderer->egl.disp, NULL, NULL) == EGL_FALSE) {
+		wlr_log(L_ERROR, "Failed to initialize EGL: %s", egl_error());
+		goto error_egl;
+	}
+
+	if (!egl_get_config(renderer->egl.disp, &renderer->egl.conf)) {
+		wlr_log(L_ERROR, "Failed to get EGL config");
+		goto error_egl;
+	}
+
+	static const EGLint attribs[] = {EGL_CONTEXT_CLIENT_VERSION, 3, EGL_NONE};
+
+	renderer->egl.context = eglCreateContext(renderer->egl.disp,
+			renderer->egl.conf, EGL_NO_CONTEXT, attribs);
+
+	if (renderer->egl.context == EGL_NO_CONTEXT) {
+		wlr_log(L_ERROR, "Failed to create EGL context: %s", egl_error());
+		goto error_egl;
+	}
+
+	backend->event_src.drm = wl_event_loop_add_fd(backend->event_loop,
+			backend->fd, WL_EVENT_READABLE, drm_event, NULL);
+	if (!backend->event_src.drm) {
+		wlr_log(L_ERROR, "Failed to create DRM event source");
+		goto error_egl;
+	}
+
+	renderer->fd = fd;
+	renderer->backend = backend;
+
+	return true;
+
+error_egl:
+	eglTerminate(renderer->egl.disp);
+	eglReleaseThread();
+	eglMakeCurrent(EGL_NO_DISPLAY, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
+
+error_gbm:
+	gbm_device_destroy(renderer->gbm);
+
+	return false;
+}
+
+void wlr_drm_renderer_free(struct wlr_drm_renderer *renderer)
+{
+	if (!renderer)
+		return;
+
+	eglDestroyContext(renderer->egl.disp, renderer->egl.context);
+	eglTerminate(renderer->egl.disp);
+	eglReleaseThread();
+	eglMakeCurrent(EGL_NO_DISPLAY, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
+
+	gbm_device_destroy(renderer->gbm);
+}
+
+static int find_id(const void *item, const void *cmp_to)
+{
+	const struct wlr_drm_display *disp = item;
+	const uint32_t *id = cmp_to;
+
+	if (disp->connector < *id)
+		return -1;
+	else if (disp->connector > *id)
+		return 1;
+	else
+		return 0;
+}
+
+void wlr_drm_scan_connectors(struct wlr_drm_backend *backend)
+{
+	drmModeRes *res = drmModeGetResources(backend->fd);
+	if (!res) {
+		wlr_log(L_ERROR, "Failed to get DRM resources");
+		return;
+	}
+
+	for (int i = 0; i < res->count_connectors; ++i) {
+		uint32_t id = res->connectors[i];
+
+		drmModeConnector *conn = drmModeGetConnector(backend->fd, id);
+		if (!conn) {
+			wlr_log(L_ERROR, "Failed to get DRM connector");
+			continue;
+		}
+
+		struct wlr_drm_display *disp;
+		int index = list_seq_find(backend->displays, find_id, &id);
+
+		if (index == -1) {
+			disp = calloc(1, sizeof *disp);
+			if (!disp) {
+				wlr_log(L_ERROR, "Allocation failed: %s", strerror(errno));
+				drmModeFreeConnector(conn);
+				continue;
+			}
+
+			disp->renderer = &backend->renderer;
+			disp->state = DRM_DISP_DISCONNECTED;
+			disp->connector = res->connectors[i];
+			snprintf(disp->name, sizeof disp->name, "%s-%"PRIu32,
+				 conn_name[conn->connector_type],
+				 conn->connector_type_id);
+
+			list_add(backend->displays, disp);
+			wlr_log(L_INFO, "Found display '%s'", disp->name);
```

Comment:
Something I'd like to do here is have the backend register the wl_output globals in the display's registry, and have ownership over those resources.